### PR TITLE
Paper: Codebraid Preview for VS Code: Pandoc Markdown Preview with Jupyter Kernels

### PR DESCRIPTION
**Pull request by gpoore**
If you are creating this PR in order to submit a draft of your paper,
see http://procbuild.scipy.org/ for logs generated by the build
process.

See the [project readme](https://github.com/scipy-conference/scipy_proceedings#build-your-paper)
for more information.

*Editor*: @mepa

*Reviewers*: @mwcraig

### CODE REVIEW

**Comment by scoobies** [comment]
Hi!, I'm @scoobies, a friendly bot.

Type ```@scoobies help``` to discover how I can help you.

**Comment by gpoore** [comment]
Just a note for anyone who may read this and then try the software over the weekend: I am a few days behind schedule, so the non-blocking code execution features described in the paper are not yet in the released versions of the software. I will provide an update on availability at the beginning of next week.

**Comment by deniederhut** [comment]
@scoobies mark paper

**Comment by deniederhut** [comment]
@scoobies build status

**Comment by scoobies** [comment]
{"gpoore-2022"=>{"data"=>{"build_output"=>"[*] Updating proceedings build tools in {self.master_repo_path}...\nAlready up to date.\n[*] Check out paper repository...\nCloning into '/tmp/tmpv2l9lvtk'...\nMoving proceedings build tools to temp directory...\n[*] Build the paper...\nBuilding: geoffrey_poore\n*** Warning: /tmp/tmpv2l9lvtk/publisher/../output/geoffrey_poore/paper_stats.json does not exist.\n", "build_pdf_path"=>"/procbuild/procbuild/../cache/gpoore-2022.pdf", "build_status"=>"success", "build_timestamp"=>"28/05 03:44"}, "status"=>"success"}}

**Comment by deniederhut** [comment]
@scoobies mark needs review

**Comment by deniederhut** [comment]
@scoobies assign @mepa as editor

**Comment by scoobies** [comment]
Assigned! @mepa is now the editor

**Comment by deniederhut** [comment]
@scoobies add @mwcraig to reviewers

**Comment by scoobies** [comment]
@mwcraig added to the reviewers list!

**Comment by gpoore** [comment]
The released software now includes all described features.

**Comment by deniederhut** [comment]
Hi! We've moved the deadline for first reviews back one week, to June 16th, to give us a little more time to find reviewers for all the papers this year. You can see the updated full timeline at https://github.com/scipy-conference/scipy_proceedings#timeline-for-2022.

**Comment by mwcraig** [comment]
Hi @gpoore 👋

I hope to have a review done by the end of the day Friday. 

An installation question for you in the meantime:

I've installed VS Code and your extension, and told VSCode to use, as its Python interpreter, a conda environment called braid` in which `codebraid` is installed. 

I've installed `pandoc` via homebrew.

When I try the side-by-side display I get an error in VSCode:

```
Codebraid is not installed or is out of date (need v0.7.0+). 
Install or upgrade (https://pypi.org/project/codebraid/), then reload the extension. 
Code execution and loading output from cache is disabled.
```

How should I fix that? (synchronised scrolling works great even without `codebraid` running 😀)

**Comment by gpoore** [comment]
@mwcraig Thanks for reviewing the paper!

For the installation issues:

Are you using OS X (since you mentioned homebrew)? I think there may be an issue with the version check somehow depending on operating system, maybe something to do with VS Code needing to pass Python information to the extension or some sort of PATH issues. I got a report earlier today of the same issue under Windows with VS Code and Codebraid installed in WSL. I'll be trying to figure this out tomorrow.

**Comment by mwcraig** [comment]
> Are you using OS X (since you mentioned homebrew)? 

Yes, I'm using OSX Monterey on an M1 mac using `mambaforge` for my python.

**Comment by gpoore** [comment]
@mwcraig It looks like subprocesses launched in VS Code extensions don't inherit conda environments, so that can prevent executables from being found. I'm working on a new release to fix this...`conda run` should do what is needed.

**Comment by gpoore** [comment]
@mwcraig I've just released a new version of the VS Code extension that should fix the issues with conda environments. If you set a Python interpreter in VS Code, then its Python installation is now checked for a `codebraid` executable. If a `codebraid` executable is found and is part of a conda environment, then `conda run` is used to run `codebraid` so that the environment is activated.

The new extension version requires Codebraid v0.8.0, which I just released, so you'll need to upgrade that as well.

**Comment by deniederhut** [comment]
Hey @mwcraig ! How is the review coming along?

File: papers/geoffrey_poore/poore.md [hunk 1]
```diff
@@ -0,0 +1,451 @@
...
+another editor is used.` [#rstudio_visual]_`{=rst}  While this approach resolves the
+inaccuracy issues of a basic Markdown preview, it also gives up features such
+as scroll sync that tightly integrate the Markdown source with the preview.
+In the case of executable code, there is the additional issue of a time delay
+in rendering the preview.  Pandoc itself can typically convert even a
+relatively long document in under one second.  However, when code is executed
+as part of the document build process, preview update is blocked until code
+execution completes.
+
+```{=rst}
```
Comment (mwcraig) [inline papers/geoffrey_poore/poore.md:96]: These blocks do not render in the preview -- that might be expected, but I wanted to mention it.

Comment (mwcraig) [reply papers/geoffrey_poore/poore.md:96]: It could be because I don't have code execution working yet...

Comment (mwcraig) [reply papers/geoffrey_poore/poore.md:96]: ...or perhaps because they are references?

**Review by mwcraig — APPROVED** [review]
I apologize for the lack of detailed wording suggestions or edits -- I've read it a few times now  and the text is in excellent shape.

I have a few suggestions:

- Include a short example of multi-file configuration like you describe [here](https://github.com/gpoore/vscode-codebraid-preview/blob/master/README.md?plain=1#L50). That is a really neat feature and it is better described in the README than in the current text.
- I still had issues with the extension finding codebraid in the conda environment I'd selected in VS Code -- that isn't a paper issue, per se, so I'll open an issue on the codebraid-preview repo about it.
- Maybe related, I also couldn't get matplotlib graphs to display -- but again, that is a code issue, not a paper issue.

To summarize:

+ This extension is pretty freakin' awesome -- the combination of live preview with rendering to the final document (e.g. PDF) is wonderful.
+ There are a couple minor issues with the extension (or with my python installation) that will, I'm sure, be resolved by the time of the conference.
+ Though it is ready to publish as-is, adding more about the multi-file scrolling would enhance the paper.

**Comment by deniederhut** [comment]
Thanks Matt!

**Comment by deniederhut** [comment]
@scoobies mark pending comment

**Comment by gpoore** [comment]
@mwcraig Thanks for reviewing the paper!

Over the next few days, I will add more details about multi-file documents. I will also do some final proofreading.

@mwcraig has opened an issue on the codebraid-preview repo regarding the code issues mentioned above, so hopefully we'll be able to get that resolved soon as well.

File: papers/geoffrey_poore/poore.md (hunk 1 above)
Comment (gpoore) [reply papers/geoffrey_poore/poore.md:96]: I overlooked responding to this the other day. In Pandoc Markdown variants, the `{=<format>}` attribute indicates content that is passed raw into the output if the output has that format. So in this case, it is text that passes directly into `rst` output for the conference paper. Since the preview is HTML, this shouldn't appear.

This does raise a good question about how to write with an HTML preview if the ultimate target is another format. I've made a note to look into adding an option that will indicate in the HTML preview the presence of raw content in formats other than HTML. It should be straightforward to make raw content optionally visible as verbatim code, with some sort of indicator to show that this is raw content.

Comment (mwcraig) [reply papers/geoffrey_poore/poore.md:96]: Thanks for the explanation!

**Comment by deniederhut** [comment]
Hi @gpoore ! How are the revisions coming along?

**Comment by gpoore** [comment]
@deniederhut I just released a new version of the software that should fix all remaining code issues that @mwcraig found, and am now working on finishing the paper. That will be finished by tomorrow night at the latest.

**Comment by mwcraig** [comment]
The new text looks good and more fully explains multi-file scrolling. 

As @gpoore indicated, all the issues I found have been addressed.

@deniederhut -- This is ready to publish as far as I'm concerned 😀

**Comment by gpoore** [comment]
I've finished my proofreading and revisions.

@mwcraig, thanks again for your help with this! And for your help finding and fixing code issues!

@deniederhut, I believe this is ready to publish.

**Comment by mepa** [comment]
@scoobies check references

**Comment by scoobies** [comment]
```
Reference check summary (note 'MISSING' DOIs are suggestions that need verification):

OK DOIs

- 10.3233/978-1-61499-649-1-87 is OK
- 10.1109/MCSE.2021.3059263 is OK
- 10.25080/Majora-7ddc1dd1-008 is OK

MISSING DOIs

- None

INVALID DOIs

- None
```

**Comment by mepa** [comment]
@scoobies mark ready
